pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,197 @@
+const nodeIndexField = "spec.nodeName"
+const parentAppLabel = "parent.app"
+
+type CniNodeTaintReconciler struct {
+	kube_client.Client
+	Log logr.Logger
+}
+
+func (r *CniNodeTaintReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("node", req.NamespacedName)
+	log.V(1).Info("event received")
+
+	// Fetch the Node instance
+	node := &kube_core.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			log.Error(err, "node not found")
+			return kube_ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch node")
+		return kube_ctrl.Result{}, err
+	}
+	log.Info("node successfully fetched")
+
+	kubeSystemPods := &kube_core.PodList{}
+	namespaceOption := kube_client.InNamespace("kube-system")
+	matchingFields := kube_client.MatchingFields{nodeIndexField: node.Name}
+	matchingLabels := kube_client.MatchingLabels{parentAppLabel: metadata.KumaCNI}
+	if err := r.Client.List(ctx, kubeSystemPods, namespaceOption, matchingFields, matchingLabels); err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err := r.updateTaints(ctx, log, node, kubeSystemPods.Items)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update node taints")
+	}
+
+	return kube_ctrl.Result{}, err
+}
+
+func (r *CniNodeTaintReconciler) updateTaints(ctx context.Context, log logr.Logger, node *kube_core.Node, pods []kube_core.Pod) error {
+	if hasTaint(node) {
+		if hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and taint")
+			return r.untaintNode(ctx, log, node)
+		} else {
+			log.Info("has no cni pod running and taint")
+			return nil
+		}
+	} else {
+		if hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and no taint")
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	taintIndex := slices.IndexFunc(node.Spec.Taints, func(taint kube_core.Taint) bool {
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		node.Spec.Taints = removeTaint(node.Spec.Taints, taintIndex)
+	}
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("removed the taint from node")
+	}
+	return err
+}
+
+func removeTaint(s []kube_core.Taint, index int) []kube_core.Taint {
+	return append(s[:index], s[index+1:]...)
+}
+
+func (r *CniNodeTaintReconciler) taintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	node.Spec.Taints = append(node.Spec.Taints, kube_core.Taint{
+		Key:    nodeReadinessTaintKey,
+		Effect: kube_core.TaintEffectNoSchedule,
+	})
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("added taint to node")
+	}
+	return err
+}
+
+func hasTaint(node *kube_core.Node) bool {
```

Comment:
Same this function if it's inlined and stored in a `hasTaint` variable it would be easier to read than jumping back and forth between the 2 methods.

Reply:
I don't mind it being inlined but this is the same story as above, I can use `slices.Contains` or `slices.IndexFunc`